### ISSUE TITLE
selectOnClose stack overflow

### ISSUE THREAD

**Issue by timmclaurin** [select2/select2#3169]
When I have SelectOnClose = true; I am recieving this error: I am using the most recent version of 4rc2.

```
Uncaught SyntaxError: Invalid regular expression: /(^|\.)select2\.(?:.*\.|)undefined(\.|$)/: Stack overflow		jQuery 
b.event.remove		jQuery 
(anonymous function)		jQuery 
b.extend.each		jQuery 
b.fn.b.each		jQuery 
b.fn.extend.off		jSelect2.
S2.define.AttachBody._detachPositioningHandler		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.CloseOnSelect._selectTriggered		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
(anonymous function)		jQuery 
b.event.dispatch		jQuery 
b.event.add.v.handle		jQuery 
b.event.trigger		jQuery 
(anonymous function)		jQuery 
b.extend.each		jQuery 
b.fn.b.each		jQuery 
b.fn.extend.trigger		jSelect2.
S2.define.SelectOnClose._handleSelectOnClose		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.CloseOnSelect._selectTriggered		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
(anonymous function)		jQuery 
b.event.dispatch		jQuery 
b.event.add.v.handle		jQuery 
b.event.trigger		jQuery 
(anonymous function)		jQuery 
b.extend.each		jQuery 
b.fn.b.each		jQuery 
b.fn.extend.trigger		jSelect2.
S2.define.SelectOnClose._handleSelectOnClose		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.CloseOnSelect._selectTriggered		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
(anonymous function)		jQuery 
b.event.dispatch		jQuery 
b.event.add.v.handle		jQuery 
b.event.trigger		jQuery 
(anonymous function)		jQuery 
b.extend.each		jQuery 
b.fn.b.each		jQuery 
b.fn.extend.trigger		jSelect2.
S2.define.SelectOnClose._handleSelectOnClose		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.CloseOnSelect._selectTriggered		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
(anonymous function)		jQuery 
b.event.dispatch		jQuery 
b.event.add.v.handle		jQuery 
b.event.trigger		jQuery 
(anonymous function)		jQuery 
b.extend.each		jQuery 
b.fn.b.each		jQuery 
b.fn.extend.trigger		jSelect2.
S2.define.SelectOnClose._handleSelectOnClose		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.CloseOnSelect._selectTriggered		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
(anonymous function)		jQuery 
b.event.dispatch		jQuery 
b.event.add.v.handle		jQuery 
b.event.trigger		jQuery 
(anonymous function)		jQuery 
b.extend.each		jQuery 
b.fn.b.each		jQuery 
b.fn.extend.trigger		jSelect2.
S2.define.SelectOnClose._handleSelectOnClose		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.CloseOnSelect._selectTriggered		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
(anonymous function)		jQuery 
b.event.dispatch		jQuery 
b.event.add.v.handle		jQuery 
b.event.trigger		jQuery 
(anonymous function)		jQuery 
b.extend.each		jQuery 
b.fn.b.each		jQuery 
b.fn.extend.trigger		jSelect2.
S2.define.SelectOnClose._handleSelectOnClose		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.CloseOnSelect._selectTriggered		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
(anonymous function)		jQuery 
b.event.dispatch		jQuery 
b.event.add.v.handle		jQuery 
b.event.trigger		jQuery 
(anonymous function)		jQuery 
b.extend.each		jQuery 
b.fn.b.each		jQuery 
b.fn.extend.trigger		jSelect2.
S2.define.SelectOnClose._handleSelectOnClose		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.CloseOnSelect._selectTriggered		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke
```

**Comment by timmclaurin** [comment]
Is this the appropriate information for a bug?

**Comment by kevin-brown** [comment]
I still need to get around to reproducing and fixing the issue, but I think this is enough information to track it down.

**Comment by timmclaurin** [comment]
OK, NP, just first bug :+1:

**Comment by timmclaurin** [comment]
Seems to not be resolved. I have updated to 4.0.0.0 and now receive:
Uncaught RangeError: Maximum call stack size exceeded

Using jQuery 19.1

**Comment by gmitra** [comment]
The issue occurs if you have the options "selectOnClose" and "closeOnSelect" (true by default) both set to true.

closeOnSelect will fire "close" which triggers selectOnClose which in turn will fire "select" which triggers closeOnSelect causing an infinite loop.

**Comment by kevin-brown** [comment]
Looks like we should be triggering a warning when we detect both `selectOnClose` and `closeOnSelect` set to `true`.

**Comment by gmitra** [comment]
I wanted both behaviors so I added a "suppressClose" flag to the "select" trigger in selectOnClose.  In closeOnSelect, I check the flag and do nothing if it's set to true.

```
  SelectOnClose.prototype._handleSelectOnClose = function () {
    var $highlightedResults = this.getHighlightedResults();

    if ($highlightedResults.length < 1) {
      return;
    }

    this.trigger('select', {
        // Added flag to prevent inifite loop of close and select 
        // events being fired when both selectOnClose and closeOnSelect
        // options are set to true
        suppressClose: true,
        data: $highlightedResults.data('data')
    });
  };
```
```
  CloseOnSelect.prototype.bind = function (decorated, container, $container) {
    var self = this;

    decorated.call(this, container, $container);

    container.on('select', function (evt) {
        // Added flag to prevent inifite loop of close and select 
        // events being fired when both selectOnClose and closeOnSelect
        // options are set to true
        if (evt && evt.suppressClose === true) {
            return;
        }
      self._selectTriggered(evt);
    });

    container.on('unselect', function (evt) {
      self._selectTriggered(evt);
    });
  };
```

**Comment by JamesKhoury** [comment]
I worked around this a different way

    container.on('select', function (evt) {
        window.setTimeout(function ()  {
            if (container.isOpen()) {
                self._selectTriggered(evt);
            }
        }, 1);
    });

**Comment by Synchro** [comment]
I'm seeing this with current `dist/js/select2.js` even though I don't have `selectOnClose` and `closeOnSelect` enabled at the same time -  I'm using only default settings, i.e. `selectOnClose: false` and `closeOnSelect: true`.

**Comment by ianawilson** [comment]
Running into this issue too. Would a solution like what @gmitra suggested be acceptable? I'm musing on the idea of contributing a PR to help move this forward

**Comment by niels-s** [comment]
Is there any update on this issues? I've tried @ianawilson / @gmitra but I've noticed the behaviour isn't the same. When you configure `closeOnSelect: true`, select an option, the next input is focussed. With the approach of @ianawilson / @gmitra when configuring both `closeOnSelect` and `selectOnClose` on `true` and applying the fix when you select an option (or close it) it doesn't focus the next input field. I'm not sure if that's expected behaviour or not? I'm not so familiar with the project so maybe anybody more familiar can confirm?
